File: cpp/linkedlist/singly_linked_list.cpp
#include <iostream>
#include <memory>
#include <stdexcept>

/*
Односвязный список.

Добавление/удаление в начало: O(1).

Добавление в конец: O(1)

Удаление из конца не реализуется в данной структуре, ввиду того, что это будет стоить O(n)
на каждый запрос.

Доступ к случайному элементу: O(n)
*/


template<typename T>
class SinglyLinkedList {
private:
    struct ListNode {
        T data;
        std::shared_ptr<ListNode> next;

        explicit ListNode(const T &value) : data(value) {}
    };
    std::shared_ptr<ListNode> head;
    std::weak_ptr<ListNode> tail;
    size_t size = 0;

public:
    void push_back(const T& value) {
        auto new_node = std::make_shared<ListNode>(value);
        new_node->data = value;
        if (!head) {
            head = new_node;
            tail = new_node;
        }
        else {
            tail.lock()->next = new_node;
            tail = new_node;
        }
        size++;

    }
    void push_front(const T& value) {
        auto new_node = std::make_shared<ListNode>(value);
        new_node->data = value;
        if (!head) {
            head = new_node;
            tail = new_node;
        }
        else {
            new_node->next = head;
            head = new_node;
        }
        size++;
    }
    void pop_front() {
        if (!head) {
            return;
        }
        head = std::move(head->next); // удалятся старый head
        if (!head) {
            tail.reset();
        }
        size--;
    }
    T& front() {
        if (!head) {
            throw std::out_of_range("list is empty");
        }
        return head->data;
    }
    T& back() {
        if (auto shared_tail = tail.lock()) {
            return shared_tail->data;
        }
        throw std::out_of_range("list is empty");
    }
    void print() const {
        auto current = head;
        while (current) {
            std::cout << current->data << " ";
            current = current->next;
        }
        std::cout << std::endl;
    }
    T& get_val(size_t index) {
        if (index >= size) {
            throw std::out_of_range("Index out of bounds");
        }
        auto current = head;
        while (current && index > 0) {
            index--;
            current = current->next;
        }
        return current->data;
    }
    [[nodiscard]] size_t get_size() const {
        return size;
    }

};

int main() {
     // тесты
    SinglyLinkedList<int> list;

    for (int i = 0; i < 6; i++) {
        list.push_back(i);
    }

    list.print();
    list.pop_front();
    list.print();

    list.push_front(-10);
    list.push_front(100);
    list.print();

    std::cout << list.front() << std::endl;
    std::cout << list.back() << std::endl;
    std::cout << list.get_size() << std::endl;

    std::cout<<list.get_val(5)<<std::endl;

    return 0;
}

File: python/linked_list/singly_linked_list.py
class SingleListNode:
    def __init__(self, value):
        self.value = value
        self.next: [SingleListNode, None] = None


class SinglyLinkedList:
    def __init__(self):
        self.__head: [SingleListNode, None] = None
        self.__tail: [SingleListNode, None] = None
        self.__size: int = 0

    def push_back(self, value):
        new_node = SingleListNode(value)

        if not self.__head:
            self.__head = new_node
        else:
            self.__tail.next = new_node

        self.__tail = new_node
        self.__size += 1

    def push_front(self, value):
        new_node = SingleListNode(value)
        if not self.__tail:
            self.__tail = new_node
        else:
            new_node.next = self.__head

        self.__head = new_node

    def pop_front(self):
        pass
        self.__size += 1

    def print_elements(self):
        """
        Это не Python стиль, но я должен показать алгоритмический смысл
        """
        current = self.__head
        print("{", end="")
        while current:
            ending = ", " if current.next else ""
            print(current.value, end=ending)
            current = current.next
        print("}")

    def insert(self, index: int, value):
        if not (0 <= index < self.__size):
            raise IndexError("Ошибка: индекс вне диапазона")

        if index == 0:
            self.push_front(value)
            return
        if index == self.__size - 1:
            self.push_back(value)
            return

        new_node = SingleListNode(value)
        current = self.__head
        for _ in range(index-1):
            current = current.next

        new_node.next = current.next
        current.next = new_node
        self.__size += 1

    def remove(self, index: int):
        if not (0 <= index < self.__size):
            raise IndexError("Ошибка: индекс вне диапазона")

        if index == 0:
            self.pop_front()
            return
        current = self.__head
        for _ in range(index-1):
            current = current.next

        current.next = current.next.next
        self.__size -= 1


    def head(self):
        if not self.__head:
            raise IndexError("Обращение к пустому списку")

        return self.__head.value

    def tail(self):
        if not self.__tail:
            raise IndexError("Обращение к пустому списку")
        return self.__tail.value

    @property
    def size(self):
        return self.__size

